### DISCUSSION TITLE
Apply failed with 1 conflict

### DISCUSSION THREAD

**Discussion by plaffitt** [kubernetes-sigs/kubebuilder#2304] [Support] [answered]
Hello,

I'm experiencing an error since a few days and I can't find the root cause of it neither any satisfying explanations about it. Here is the error message:

> 2021-08-19T09:48:16.090Z    ERROR    controller-runtime.manager.controller.pod    Reconciler error    {"reconciler group": "", "reconciler kind": "Pod", "name": "nginx-deployment-7778499998-czhxm", "namespace": "default", "error": "**Apply failed with 1 conflict: conflict with \"manager\" using dcr.enix.io/v1alpha1: .status**"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
    /go/pkg/mod/k8s.io/apimachinery@v0.20.6/pkg/util/wait/wait.go:99

I don't understand where this "manager" comes from and what does it means by conflict? I mean, I got this error at the creation of a CR, how can it conflicts with nothing? And why the status is conflicting while I don't even set it?

Here is a sample of my code (It seems that the errors come from the `err = r.Patch(ctx, &cachedImage, client.Apply, applyOpts...)` line):
```golang
func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	var pod corev1.Pod
	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
		return ctrl.Result{}, client.IgnoreNotFound(err)
	}

	cachedImages, err := r.desiredCachedImages(pod)
	if err != nil {
		return ctrl.Result{}, err
	}

	applyOpts := []client.PatchOption{
		client.FieldOwner("pod-controller"),
	}

	for _, cachedImage := range cachedImages {
		_ = cachedImage
		err = r.Patch(ctx, &cachedImage, client.Apply, applyOpts...)
		if err != nil {
			return ctrl.Result{}, err
		}
	}

	return ctrl.Result{}, nil
}

func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&corev1.Pod{}).
		Complete(r)
}

func (r *PodReconciler) desiredCachedImages(pod corev1.Pod) ([]dcrenixiov1alpha1.CachedImage, error) {
	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
	cachedImages := []dcrenixiov1alpha1.CachedImage{}

	for i, container := range containers {
		sourceImage, ok := pod.Annotations[fmt.Sprintf("tugger-original-image-%d", i)]
		if !ok {
			continue
		}
		re := regexp.MustCompile(`localhost:[0-9]+/`)
		image := string(re.ReplaceAll([]byte(container.Image), []byte("")))
		sanitizedName := cache.SanitizeImageName(image)
		cachedImage := dcrenixiov1alpha1.CachedImage{
			TypeMeta: metav1.TypeMeta{APIVersion: dcrenixiov1alpha1.GroupVersion.String(), Kind: "CachedImage"},
			ObjectMeta: metav1.ObjectMeta{
				Name: sanitizedName,
			},
			Spec: dcrenixiov1alpha1.CachedImageSpec{
				Image:       image,
				SourceImage: sourceImage,
			},
		}
		cachedImages = append(cachedImages, cachedImage)
	}

	return cachedImages, nil
}
```

EDIT: I surprisingly "fixed" my issue by just adding `client.ForceOwnership` to my `PatchOptions`. I don't get the error anymore, but I still don't understand what is going on exactly. I feel like I added `!important` in some css stylesheet.

**Comment by camilamacedo86** [comment, marked as answer]
Your error indicates that there's a conflict during a server-side apply operation, specifically related to the .status field of a resource managed by another controller named "manager". This conflict arises due to overlapping ownership of fields when using server-side apply in Kubernetes.

Let's dissect this:

- Server-side Apply and Field Management: Server-side apply uses a mechanism called field management. Every field in a resource is owned by a manager (a controller, user, etc.) that has set its value using the server-side apply. The manager's name is typically derived from the User-Agent HTTP header, but with controller-runtime, it's often the controller's name. In your error, there seems to be a conflict with another manager named "manager".

- Conflict with .status Field: Even though you didn't directly change the .status field, you might be changing something that indirectly affects this field, or the server-side apply might consider that you're trying to claim ownership of it.

- client.ForceOwnership: By adding client.ForceOwnership to the PatchOptions, you're essentially saying: "I don't care about the previous owners of these fields; I claim ownership now." This is a way to force your way through the conflict, but be careful. It can lead to scenarios where you overwrite changes unintentionally, especially if multiple controllers are managing the same resources.

However, server-side is not support yet in controller-tools or controller-runtime so consequently it is not support in kubebuilder too, You can check the progress of this work here: 
- https://github.com/kubernetes-sigs/controller-runtime/issues/347
- https://github.com/kubernetes-sigs/controller-tools/pull/536

Marking it as answered and closing this one.
However, feel free to re-open if you see that fits.
